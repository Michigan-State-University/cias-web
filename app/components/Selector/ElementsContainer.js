import React from 'react';
import PropTypes from 'prop-types';

import decideIfPassValue from 'utils/decideIfPassValue';
import Column from 'components/Column';
import Text from 'components/Text';
import { themeColors, colors, boxShadows } from 'theme';

const ElementsContainer = ({
  options,
  selectedOption,
  setOption,
  toggleActive,
}) => {
  const handleSelectOption = id => {
    toggleActive();
    setOption(id);
  };

  return (
    <Column
      bg={colors.white}
      shadow={boxShadows.selago}
      px={20}
      py={15}
      align="start"
    >
      {options.map(({ label, id }, index) => {
        const isSelected = id === selectedOption.id;
        return (
          <Text
            key={`el-text-option-${id}`}
            color={isSelected ? themeColors.secondary : null}
            mb={decideIfPassValue({
              index,
              arrayLength: options.length,
              value: 10,
            })}
            fontWeight="bold"
            whiteSpace="nowrap"
            clickable
            onClick={() => handleSelectOption(id)}
            role="option"
          >
            {label}
          </Text>
        );
      })}
    </Column>
  );
};

ElementsContainer.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({ id: PropTypes.string, label: PropTypes.string }),
  ),
  setOption: PropTypes.func,
  toggleActive: PropTypes.func,
  selectedOption: PropTypes.shape({
    id: PropTypes.string,
    label: PropTypes.string,
  }),
};

export default ElementsContainer;
