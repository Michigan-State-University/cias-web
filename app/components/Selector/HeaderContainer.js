import React from 'react';
import PropTypes from 'prop-types';

import Text from 'components/Text';
import { themeColors } from 'theme';

const HeaderContainer = ({ children, onClick, disabled }) => (
  <Text
    color={themeColors.secondary}
    fontWeight="bold"
    clickable
    disabled={disabled}
    onClick={() => !disabled && onClick()}
  >
    {children}
  </Text>
);

HeaderContainer.propTypes = {
  children: PropTypes.node,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
};

export default HeaderContainer;
