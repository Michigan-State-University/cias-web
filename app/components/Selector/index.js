import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';

import Tooltip from 'components/Tooltip';
import Box from 'components/Box';
import Row from 'components/Row';
import questionMark from 'assets/svg/question-mark.svg';

import useOutsideClick from 'utils/useOutsideClick';
import ToggleArrow from 'components/ToggleArrow';
import HeaderContainer from './HeaderContainer';
import ElementsContainer from './ElementsContainer';

const Selector = ({
  options,
  activeOption,
  tooltipContent,
  ElementsComponent,
  HeaderComponent,
  rightPosition,
  setOption,
  selectOptionPlaceholder,
  disabled,
  ariaLabel,
}) => {
  const selector = useRef(null);

  const selectedOption = activeOption || { label: selectOptionPlaceholder };
  const [isActive, setIsActive] = useState(false);

  useOutsideClick(selector, () => setIsActive(false), isActive);

  const toggleActive = () => setIsActive(!isActive);

  return (
    <Box position="relative" ref={selector}>
      <Row align="center" role="button">
        <HeaderComponent disabled={disabled} onClick={toggleActive}>
          {selectedOption.label}
        </HeaderComponent>
        {tooltipContent && (
          <Tooltip
            id={`el-tooltip-id-${selectedOption.id}`}
            ml={8}
            icon={questionMark}
            text={tooltipContent}
          />
        )}
        {!disabled && (
          <ToggleArrow
            facingUp={isActive}
            clickable
            ml={8}
            onClick={toggleActive}
          />
        )}
      </Row>
      {isActive && (
        <Row
          position="absolute"
          top="30px"
          right={rightPosition || '0'}
          zIndex={999}
          role="listbox"
          aria-label={ariaLabel}
        >
          <ElementsComponent
            options={options}
            selectedOption={selectedOption}
            setOption={setOption}
            toggleActive={toggleActive}
          />
        </Row>
      )}
    </Box>
  );
};

Selector.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({ id: PropTypes.string, label: PropTypes.string }),
  ),
  activeOption: PropTypes.shape({
    id: PropTypes.string,
    label: PropTypes.string,
  }),
  tooltipContent: PropTypes.string,
  HeaderComponent: PropTypes.elementType,
  ElementsComponent: PropTypes.elementType,
  rightPosition: PropTypes.string,
  setOption: PropTypes.func,
  selectOptionPlaceholder: PropTypes.string,
  disabled: PropTypes.bool,
  ariaLabel: PropTypes.string,
};

Selector.defaultProps = {
  HeaderComponent: HeaderContainer,
  ElementsComponent: ElementsContainer,
};
export default Selector;
