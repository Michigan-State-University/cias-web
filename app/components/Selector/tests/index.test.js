/**
 *
 * Tests for Selector
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import 'jest-styled-components';

import Selector from '../index';

const defaultProps = {
  options: [
    { label: 'Label', id: 'asd02as' },
    { label: 'Label2', id: 'as0-12ds' },
  ],
  activeOption: { label: 'Label', id: 'asd02as' },
  tooltipContent: 'tooltipContent',
};

describe('<Selector />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<Selector {...defaultProps} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const { container } = render(<Selector {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
