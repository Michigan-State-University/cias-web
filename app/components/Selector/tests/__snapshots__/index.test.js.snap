// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Selector /> Should render and match the snapshot 1`] = `
.c0 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  position: relative;
}

.c3 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: 8px;
}

.c4 {
  width: auto;
  height: auto;
}

.c6 {
  width: auto;
  height: auto;
  margin-left: 8px;
  cursor: pointer;
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c2 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  font-weight: 700;
  color: rgba(200,102,234,1);
  cursor: pointer;
  color: rgba(200,102,234,1);
  cursor: pointer;
}

.c5 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  color: #2F3850;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <p
        class="c2"
        color="#C866EA"
        font-weight="bold"
      >
        Label
      </p>
      <div
        class="c3"
        display="flex"
      >
        <img
          alt="?"
          class="c4"
          currentitem="false"
          data-for="el-tooltip-id-asd02as"
          data-tip="true"
          src="IMAGE_MOCK"
        />
        <div
          class="__react_component_tooltip t00000000-0000-0000-0000-000000000000 place-top type-dark"
          data-id="tooltip"
          id="el-tooltip-id-asd02as"
        >
          <style>
            
  	.t00000000-0000-0000-0000-000000000000 {
	    color: #fff;
	    background: #222;
	    border: 1px solid transparent;
  	}

  	.t00000000-0000-0000-0000-000000000000.place-top {
        margin-top: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::before {
        border-top: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        bottom: -6px;
        left: 50%;
        margin-left: -8px;
        border-top-color: #222;
        border-top-style: solid;
        border-top-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-bottom {
        margin-top: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::before {
        border-bottom: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        top: -6px;
        left: 50%;
        margin-left: -8px;
        border-bottom-color: #222;
        border-bottom-style: solid;
        border-bottom-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-left {
        margin-left: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::before {
        border-left: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        right: -6px;
        top: 50%;
        margin-top: -4px;
        border-left-color: #222;
        border-left-style: solid;
        border-left-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-right {
        margin-left: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::before {
        border-right: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        left: -6px;
        top: 50%;
        margin-top: -4px;
        border-right-color: #222;
        border-right-style: solid;
        border-right-width: 6px;
    }
  
          </style>
          <p
            class="c5"
          >
            tooltipContent
          </p>
        </div>
      </div>
      <img
        alt="arrow"
        class="c6"
        src="IMAGE_MOCK"
        transform=""
      />
    </div>
  </div>
</div>
`;

exports[`<Selector /> Should render disabled and match the snapshot 1`] = `
.c0 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  position: relative;
}

.c3 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: 8px;
}

.c4 {
  width: auto;
  height: auto;
}

.c6 {
  width: auto;
  height: auto;
  margin-left: 8px;
  cursor: pointer;
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c2 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  font-weight: 700;
  color: rgba(200,102,234,1);
  cursor: pointer;
  color: rgba(200,102,234,1);
  cursor: pointer;
}

.c5 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  color: #2F3850;
}

.c9 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  margin-bottom: 10px;
  font-weight: 700;
  color: rgba(200,102,234,1);
  white-space: nowrap;
  cursor: pointer;
  color: rgba(200,102,234,1);
  cursor: pointer;
}

.c10 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  font-weight: 700;
  color: #2F3850;
  white-space: nowrap;
  cursor: pointer;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
  position: absolute;
  z-index: 999;
  top: 30px;
  right: 0;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  -webkit-align-content: flex-end;
  -ms-flex-line-pack: end;
  align-content: flex-end;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  background-color: rgba(255,255,255,1);
  box-shadow: 0px 4px 20px #E3EEFB;
  padding-top: 15px;
  padding-bottom: 15px;
  padding-right: 20px;
  padding-left: 20px;
}

@media only screen and (min-width:768px) {

}

@media only screen and (min-width:960px) {

}

@media only screen and (min-width:1170px) {

}

@media only screen and (min-width:1360px) {

}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <p
        class="c2"
        color="#C866EA"
        font-weight="bold"
      >
        Label
      </p>
      <div
        class="c3"
        display="flex"
      >
        <img
          alt="?"
          class="c4"
          currentitem="false"
          data-for="el-tooltip-id-asd02as"
          data-tip="true"
          src="IMAGE_MOCK"
        />
        <div
          class="__react_component_tooltip t00000000-0000-0000-0000-000000000000 place-top type-dark"
          data-id="tooltip"
          id="el-tooltip-id-asd02as"
        >
          <style>
            
  	.t00000000-0000-0000-0000-000000000000 {
	    color: #fff;
	    background: #222;
	    border: 1px solid transparent;
  	}

  	.t00000000-0000-0000-0000-000000000000.place-top {
        margin-top: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::before {
        border-top: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        bottom: -6px;
        left: 50%;
        margin-left: -8px;
        border-top-color: #222;
        border-top-style: solid;
        border-top-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-bottom {
        margin-top: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::before {
        border-bottom: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        top: -6px;
        left: 50%;
        margin-left: -8px;
        border-bottom-color: #222;
        border-bottom-style: solid;
        border-bottom-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-left {
        margin-left: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::before {
        border-left: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        right: -6px;
        top: 50%;
        margin-top: -4px;
        border-left-color: #222;
        border-left-style: solid;
        border-left-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-right {
        margin-left: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::before {
        border-right: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        left: -6px;
        top: 50%;
        margin-top: -4px;
        border-right-color: #222;
        border-right-style: solid;
        border-right-width: 6px;
    }
  
          </style>
          <p
            class="c5"
          >
            tooltipContent
          </p>
        </div>
      </div>
      <img
        alt="arrow"
        class="c6"
        src="IMAGE_MOCK"
        transform="rotate(180deg);"
      />
    </div>
    <div
      class="c7"
    >
      <div
        class="c8"
      >
        <p
          class="c9"
          color="#C866EA"
          font-weight="bold"
        >
          Label
        </p>
        <p
          class="c10"
          font-weight="bold"
        >
          Label2
        </p>
      </div>
    </div>
  </div>
</div>
`;
